hub/core/storage/provider.py

```diff
@@ -0,0 +1,139 @@
+from collections.abc import MutableMapping
+from typing import Optional
+from hub.core.storage.utils import check_byte_indexes
+from hub.constants import BYTE_PADDING
+
+
+class Provider(MutableMapping):
+    def __init__(self):
+        self.mapper = {}
+
+    def __getitem__(
+        self,
+        path: str,
+        start_byte: Optional[int] = None,
+        end_byte: Optional[int] = None,
+    ):
+        """
+        Gets the object present at the path. 
+        Optionally, if the start_byte and/or end_byte arguments are specified, it only returns required bytes
```

Comment:
maybe the subset (start_byte:end_byte) functionality should not be present in the provider, but rather the caching class?

because we never want to download a subset of the chunk if we can avoid it, it should download the entire chunk, cache it, & then return the start_byte:end_byte.

it would be idea that cache chain handles this.

Reply:
The provider interface deals solely with getting bytes from a file. For us, "file" is usually a chunk and "bytes" can either mean the whole thing, or a subset of the "file". It doesn't make sense in the context of chunking, but in the context of the interface it makes sense.

This will prove useful when we reuse these providers for caching, as well as things like getting the metadata. I can come up with more examples, but I think it's reasonable to keep the byte-range part of this interface, especially when it might afford us some optimizations that we would otherwise not have access to (like the S3 get object range parameter)